builder/build_test.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) OpenFaaS Project 2017. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package builder
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"testing"
+)
+
+func Test_CopyFiles(t *testing.T) {
+	fileModes := []int{0600, 0640, 0644, 0700, 0755}
+
+	dir := os.TempDir()
+	for _, mode := range fileModes {
+		// set up a source folder with 2 file
+		srcDir, srcDirErr := setupSourceFolder(2, mode)
+		if srcDirErr != nil {
+			log.Fatal("Error creating source folder")
+		}
+		defer os.RemoveAll(srcDir)
```

Comment:
You may also want to do this up front for the case where remove didn't work or is running in parallel on the same system. Can you randomise the path any more than this?

Reply:
@alexellis, the temporary folders had some random integers attached at the end of the directory name example `openfaas-test-destination-290650684`. So I guess that will handle the case where several tests are run in parallel on the same system.

However, I notice that if a Fatal is logged then the directed till that point are not deleted. So before throwing the Fatals we should clean up all the temporary directories created. I will update the PR tomorrow with the relevant changes.

Reply:
@ganesshkumar I prefer using t.Error() and call a function doing the teardown ops, as in fetch_template_test.go

Reply:
@itscaro I need to merge these changes and release to unblock @ganesshkumar and a couple of other people for the Java template. Please could you raise a PR to fix the tests how you like them after this merge?